gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,376 @@
+
+            validater(self, body)
+
+    def validate_comment_(self, body: List[str]):
+        fmt_str = body[0]
+        num_vars = fmt_str.count("%s")
+        if num_vars != len(body) - 1:
+            raise MismatchedPrintStatement(
+                "Expected {} var names in comment but received {}"
+                .format(num_vars, len(body) - 1))
+
+        for var_name in body[1:]:
+            var_base = var_name.split(".")[0]
+            if var_base not in self.var_defs_:
+                raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                                 .format(var_base))
+
+    def validate_print_(self, body: List[str]):
+        fmt_str = body[0]
+        num_prints = fmt_str.count("%s")
+        if num_prints != len(body) - 1:
+            raise MismatchedPrintStatement(
+                "Expected {} expresssions in print statement but received {}"
+                .format(num_prints, len(body) - 1))
+
+        for expression in body[1:]:
+            var = expression.split(".")[0]
+            if var not in self.var_defs_:
+                raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                                 .format(var))
+
+    def validate_define_(self, body: str):
+        m = re.match(r"^([^=]+)=([^=]+)$", body)
+        if not m:
+            raise BadAssignment("Bad assignment statement: {}", body)
+
+        var, rval = m.groups()
+        if var in RESERVED_WORDS:
+            raise ReservedVariableName(
+                "Tried to define a variable with reserved name: {}".format(var))
+
+        rval_base = rval.split(".")[0]
+        if not rval_base in self.var_defs_:
+            raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                             .format(rval_base))
+        # TODO: Need to validate the attributes of the response
+        #       based on the method return type.
+        # TODO: Need to check the defined variables
+        #       if the rhs references a non-response variable.
+        #
+        # Note: really checking for safety would be equivalent to
+        #       re-implementing the python interpreter.
+        self.var_defs_.add(var)
+
+    def validate_loop_(self, body):
+        segments = set(body.keys())
+        map_args = {self.MAP_KWORD, self.BODY_KWORD}
+
+        # Even though it's valid python to use a variable outside of the lexical
+        # scope it was defined in,
+        #
+        # i.e.
+        #   for m in molluscs:
+        #     handle(m)
+        #   print("Last mollusc: {}".format(m))
+        #
+        # is valid python, the samplegen spec requires that errors are raised
+        # if strict lexical scoping is violated.
+        previous_defs = set(self.var_defs_)
+
+        if {self.COLL_KWORD, self.VAR_KWORD, self.BODY_KWORD} == segments:
+            collection_name = body[self.COLL_KWORD].split(".")[0]
```

Comment:
The collection name is actually all of `body[COLL_KWORD]`. You're right that `[0]` must be in `var_defs_`, but all the subfields must properly resolve. Add a TODO for once you're looking at the protobufs.

Same for map

Reply:
Done.